src/peer_conn_protocol.rs

```diff
@@ -5,32 +5,75 @@ use super::server;
 use std::io;
 use std::sync::mpsc::Sender;
 use uuid::Uuid;
+use std::collections::HashMap;
+use std::net::TcpStream;
 
-pub fn send_p2p_handshake(
-    conn_id: u64,
-    peer_address: String,
-    peer_port: u16,
-    messenger: Sender<MessengerOperations>,
-) -> Result<Uuid, io::Error> {
-    let stream = server::new_connection(peer_address.clone(), peer_port)?;
-    messenger
-        .send(MessengerOperations::New(NewConnectionMessage {
+pub struct Peer {
+    pub peer_address: String,
+    pub peer_port: u16,
+    pub stream: TcpStream,
+}
+
+impl Peer {
+    pub fn send_p2p_handshake(
+        &mut self,
+        conn_id: u64,
+        messenger: Sender<MessengerOperations>,
+    ) -> Result<Uuid, io::Error> {
+        
+        messenger
+            .send(MessengerOperations::New(NewConnectionMessage {
+                conn_id,
+                socket: self.stream.try_clone().unwrap(),
+            }))
+            .unwrap();
+    
+        send_packet!(
+            messenger,
             conn_id,
-            socket: stream.try_clone().unwrap(),
-        }))
+            Packet::Handshake(Handshake {
+                protocol_version: 404,
+                server_address: self.peer_address.clone(),
+                server_port: self.peer_port,
+                next_state: 6,
+            })
+        )
         .unwrap();
+    
+        Ok(Uuid::new_v4())
+    }
+}
 
-    send_packet!(
-        messenger,
-        conn_id,
-        Packet::Handshake(Handshake {
-            protocol_version: 404,
-            server_address: peer_address.clone(),
-            server_port: peer_port,
-            next_state: 6,
-        })
-    )
-    .unwrap();
-
-    Ok(Uuid::new_v4())
+pub struct PeerManager {
+    pub peer_map: HashMap::<u64, Peer>,
 }
+
+impl PeerManager{
+    pub fn new() -> PeerManager{
+        let peer_map = HashMap::<u64, Peer>::new();
+        PeerManager{peer_map}
+    }
+
+    pub fn connect_to_peer(
+        &mut self,
+        conn_id: u64,
+        peer_address: String,
+        peer_port: u16,
+        messenger: Sender<MessengerOperations>,
+    ) -> Result<Uuid, io::Error> {
+        let stream = server::new_connection(peer_address.clone(), peer_port)?;
+        let mut peer = Peer{peer_address, peer_port, stream};
+
+        match(peer.send_p2p_handshake(conn_id, messenger)){
+            Uuid => {
+                self.peer_map.insert(conn_id, peer);
```

Comment:
If you call this method twice with the same connection ID it'll overwrite the old peer(at least I assume so- since it doesn't seem to throw an error or perform a check) - not sure if that's intentional

Reply:
Oh yes you're right I forgot to increment the id there.

Reply:
What are you incrementing? The connection ID is being passed as an argument, not sure how you'll increment that to guarantee uniqueness

Reply:
From my understanding the connection id is starting at 1 at the server and then incremented for each new connection, which means it is unique

Reply:
The server function does have that variable locally- we can pass a mutable reference to that number around the program but it doesn't feel great

Reply:
I feel like there are not many other options for that. I could create a static variable and update it but I don't think it's a common thing to do in Rust. Otherwise there has to be some kind of mutable variable passing.

Reply:
What about using uuids like we talked about before?

Reply:
I'll give it a shot

Reply:
But that means I'll have to modify the messenger, right ?

Reply:
yeah you'll have to modify everything that uses the connection id

Reply:
You can also just apply the quick fix of doing the unique-by-coincidence solution and punt on the uuid thing